{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7fe9c920",
   "metadata": {},
   "source": [
    "# Ancestry analysis on regions under sweep selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51726aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -qq malariagen_data\n",
    "import malariagen_data\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import zarr\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "ag3 = malariagen_data.Ag3()\n",
    "ag3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "446bce2b",
   "metadata": {},
   "source": [
    "## workflow to prepare SNP data and plot PCoA and UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b8b0d8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install -qq umap\n",
    "#!pip install -qq umap-learn\n",
    "\n",
    "!pip install -qq  scipy\n",
    "!pip install -qq scikit-bio\n",
    "!pip install -q umap-learn\n",
    "\n",
    "import umap.umap_ as umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bff1c61e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNP data\n",
    "from dask.diagnostics.progress import ProgressBar\n",
    "\n",
    "def get_snp_data(region, sample_query, filter):\n",
    "    print(f\"Processing SNP data for region: {region}\")\n",
    "    # Construct snp data\n",
    "    gm_snps =ag3.snp_calls(\n",
    "        region=region,\n",
    "        sample_query=sample_query,)\n",
    "\n",
    "    # construct population information\n",
    "    taxon_sample = ag3.sample_metadata(sample_query=sample_query)\n",
    "    # Define pop based on iso and taxon like GM-N_gamb, GM-U_arab etc.\n",
    "    taxon_sample['pop'] = taxon_sample['admin1_iso'] + '_' + taxon_sample['taxon']\n",
    "    taxon_sample.groupby('pop').size()\n",
    "    # kepp pop >= 10\n",
    "    taxon_sample = taxon_sample[taxon_sample['pop'].isin(taxon_sample['pop'].value_counts()[taxon_sample['pop'].value_counts() >= 10].index)]\n",
    "    # SAve taxon_sample metadata\n",
    "    taxon_sample.to_csv('taxon_sample.csv', sep=',', index=True)\n",
    "\n",
    "    # Get all sample IDs from the dataset\n",
    "    dataset_sample_ids = gm_snps['sample_id'].values  # Array of sample IDs in gm_snps\n",
    "\n",
    "    # Find indices of selected samples\n",
    "    sample_indices = np.where(np.isin(dataset_sample_ids, taxon_sample['sample_id'].values))[0]\n",
    "\n",
    "    # Select samples using indices\n",
    "    gm_snps_selected = gm_snps.isel(samples=sample_indices)\n",
    "\n",
    "    samples_selected = gm_snps_selected['sample_id'].values\n",
    "\n",
    "    # To filter the SNP dataset and warp the dataset to GT array\n",
    "    filt_val = gm_snps_selected[f\"variant_filter_pass_{filter}\"].values\n",
    "    gt_filtered = allel.GenotypeDaskArray(gm_snps_selected[\"call_genotype\"][filt_val].data)\n",
    "\n",
    "    return gt_filtered, gm_snps_selected, taxon_sample\n",
    "\n",
    "\n",
    "def snp_filtering_proccess(gt_filtered, gm_snps_selected, taxon_sample):\n",
    "  # Get all sample IDs in the SNP dataset\n",
    "  dataset_sample_ids = gm_snps_selected['sample_id'].values  # Sample IDs in gm_snps_selected\n",
    "\n",
    "  # Get indices of selected samples based on taxon_sample['sample_id']\n",
    "  pop_idx = np.where(np.isin(dataset_sample_ids, taxon_sample['sample_id'].values))[0]\n",
    "  with ProgressBar():\n",
    "    ac = gt_filtered.take(pop_idx, axis=1).count_alleles(max_allele=3).compute()\n",
    "  #filter - to keep biallelic and remove missing data\n",
    "  filter_ac = (ac.max_allele() == 1) & (ac[:, :2].min(axis=1) > 1)\n",
    "  gt_rm_miss = gt_filtered.compress(filter_ac, axis=0)\n",
    "  geno_alt = gt_rm_miss.to_n_alt()\n",
    "\n",
    "  return geno_alt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d344a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform LD pruning\n",
    "import dask.array as da\n",
    "import allel\n",
    "import numpy as np\n",
    "def plot_ld(gn, title):\n",
    "    \"\"\"\n",
    "    Compute LD between SNP and plot pairwise LD.\n",
    "    \"\"\"\n",
    "    m = allel.rogers_huff_r(gn) ** 2\n",
    "    ax = allel.plot_pairwise_ld(m)\n",
    "    ax.set_title(title)\n",
    "def ld_prune(gn, size, step, threshold=.1, n_iter=1):\n",
    "    \"\"\"\n",
    "    Perform LD pruning on genotype data.\n",
    "    \"\"\"\n",
    "    for i in range(n_iter):\n",
    "        # Convert gn to a NumPy array if it's a Dask array\n",
    "        if isinstance(gn, da.Array):\n",
    "            gn = gn.compute()\n",
    "\n",
    "        loc_unlinked = allel.locate_unlinked(gn, size=size, step=step, threshold=threshold)\n",
    "        n = np.count_nonzero(loc_unlinked)\n",
    "        n_remove = gn.shape[0] - n\n",
    "        print('iteration', i + 1, 'retaining', n, 'removing', n_remove, 'variants')\n",
    "        gn = gn.compress(loc_unlinked, axis=0)\n",
    "    return gn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5160af9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SNP data processing\n",
    "regions = [\"2R:28,430,000-28,615,000\", \"3R:28,400,000-28,600,000\", \"X:15,130,000-15,324,000\", \"3R:32,000,000-32,080,000\"]\n",
    "taxons = [\"coluzzii\", \"gambiae\", \"bissau\", \"arabiensis\", \"melas\"]\n",
    "#taxons = [\"coluzzii\", \"gambiae\", \"bissau\"]\n",
    "country = \"Gambia, The\"\n",
    "sample_query = f\"country=='{country}' and taxon in {taxons}\"\n",
    "filter = 'gamb_colu_arab'\n",
    "#filter = 'gamb_colu'\n",
    "region = regions\n",
    "gt_filtered, gm_snps_selected, taxon_sample = get_snp_data(region, sample_query, filter)\n",
    "geno_alt = snp_filtering_proccess(gt_filtered, gm_snps_selected, taxon_sample)\n",
    "\n",
    "geno_alt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302dcc88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LD perform LD pruning\n",
    "gnu = ld_prune(geno_alt, size=500, step=200, threshold=.1, n_iter=10)\n",
    "plot_ld(geno_alt[:1000], 'Figure 1. Pairwise LD before LD pruning.')\n",
    "plot_ld(gnu[:1000], 'Figure 2. Pairwise LD after LD pruning.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8278ed95",
   "metadata": {},
   "outputs": [],
   "source": [
    "gnu.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82b4500d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save genotype data\n",
    "from google.colab import files\n",
    "np.save('filtered_geno_alt.npy', geno_alt)\n",
    "np.save('filtered_geno_alt_pruned.npy', gnu)\n",
    "\n",
    "files.download('filtered_geno_alt.npy')\n",
    "files.download('filtered_geno_alt.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40e8fcc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform PCoA\n",
    "from skbio.stats.ordination import pcoa\n",
    "\n",
    "# Compute pairwise distance\n",
    "matrix_distance = allel.pairwise_distance(geno_alt, metric='cityblock')\n",
    "# Perform PCoA using scikit-bio\n",
    "pcoa_results = pcoa(matrix_distance)\n",
    "\n",
    "pcoa_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdea0911",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import altair as alt\n",
    "\n",
    "# import population info\n",
    "taxon_sample = pd.read_csv('taxon_sample.csv')\n",
    "\n",
    "#load anopheles gambiae from Gambia country\n",
    "pop_info = taxon_sample[['sample_id', 'taxon', 'pop', 'location', 'admin1_iso', 'admin1_name', 'admin2_name', 'cohort_admin1_quarter']]\n",
    "\n",
    "# show variance\n",
    "pcoa_results.proportion_explained * 100\n",
    "# save variance\n",
    "variance_explained = pcoa_results.proportion_explained * 100\n",
    "variance_explained = pd.DataFrame(variance_explained)\n",
    "variance_explained.columns = ['variance_explained']\n",
    "variance_explained.index.name = 'PC'\n",
    "variance_explained.reset_index(inplace=True)\n",
    "variance_explained.groupby('PC').sum()\n",
    "variance_explained.to_csv(f'variance_explained_all_reigion.csv', sep=',', index=True)\n",
    "variance_explained\n",
    "\n",
    "# Sort the dataframe by variance explained in descending order\n",
    "variance_explained_sorted = variance_explained.sort_values(by='variance_explained', ascending=False)\n",
    "# Select the top 10 principal components\n",
    "top_10_variance_explained = variance_explained_sorted.head(10)\n",
    "# Create a bar chart using Altair\n",
    "chart = alt.Chart(top_10_variance_explained).mark_bar().encode(\n",
    "    # Map PC to the x-axis and variance_explained to the y-axis\n",
    "    x=alt.X('PC', sort='-y', title='Principal Component'),\n",
    "    y=alt.Y('variance_explained', title='Variance Explained'),\n",
    "    # Add tooltips to display PC and variance explained when hovering\n",
    "    tooltip=['PC', 'variance_explained']\n",
    ").properties(\n",
    "    title='Variance Explained by Top 10 Principal Components' # Add a title to the chart\n",
    ")\n",
    "chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d0cf62b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# join the pcoa result with the metadata\n",
    "pcoa_results_df = pcoa_results.samples.iloc[:, 0:10]\n",
    "\n",
    "pcoa_results_df['sample_id'] = pop_info['sample_id'].values\n",
    "# merge pcoa and metadata by sample_id\n",
    "pcoa_results_df = pcoa_results_df.merge(pop_info, on='sample_id', how='left')\n",
    "pcoa_results_df\n",
    "\n",
    "# save file\n",
    "pcoa_results_df.to_csv(f'pcoa_taxon_all_region_PC1_vs_PC2_all_taxon.csv', sep=',', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b1dc66d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract population labels as list to match with geno\n",
    "pop_info['taxon'] = pop_info['taxon'].replace(\n",
    "    {\n",
    "        \"coluzzii\": \"An. coluzzii\",\n",
    "        \"gambiae\": \"An. gambiae s.s\",\n",
    "        \"bissau\": \"Bissau\",\n",
    "        \"arabiensis\": \"An. arabiensis\",\n",
    "        \"melas\": \"An. melas\"\n",
    "    }\n",
    ")\n",
    "population_labels = pop_info['taxon'].tolist()\n",
    "population_labels\n",
    "\n",
    "# Plot PCoA results with colors by population\n",
    "#plt.figure(figsize=(10, 8))\n",
    "plt.figure(figsize=(12, 10))\n",
    "\n",
    "# Create a color palette for each population\n",
    "unique_populations = np.unique(population_labels)\n",
    "unique_populations\n",
    "\n",
    "# define color palette manually\n",
    "population_color_map = {\n",
    "    'Bissau': 'purple',\n",
    "    'An. gambiae s.s': 'blue',\n",
    "    'An. coluzzii': 'orange',\n",
    "    'An. arabiensis': 'green',\n",
    "    'An. melas': 'red'\n",
    "}\n",
    "\n",
    "# Scatter plot with colors assigned by population\n",
    "x=2\n",
    "y=3\n",
    "for pop in unique_populations:\n",
    "    mask = np.array(population_labels) == pop\n",
    "    plt.scatter(\n",
    "        pcoa_results.samples[f'PC{x}'][mask],\n",
    "        pcoa_results.samples[f'PC{y}'][mask],\n",
    "        label=pop,\n",
    "        edgecolor='white',     # White border\n",
    "        s=200,\n",
    "        linewidths=1.5,\n",
    "        color=population_color_map[pop])\n",
    "# Get current legend\n",
    "legend = plt.legend(title='Population', loc='best', fontsize=21, bbox_to_anchor=(1.45, 1))\n",
    "plt.setp(legend.get_title(), fontweight='bold')\n",
    "\n",
    "# Customize individual legend text\n",
    "\n",
    "for i, text in enumerate(legend.get_texts()):\n",
    "    pop_name = text.get_text()\n",
    "    if pop_name == 'Bissau':\n",
    "        text.set_fontweight('bold')\n",
    "        text.set_fontsize(22)\n",
    "    else :\n",
    "        text.set_fontstyle('italic')\n",
    "        text.set_fontweight('bold')\n",
    "        text.set_fontsize(22)\n",
    "\n",
    "# Add title and axis labels with bold font\n",
    "plt.title(f'PCoA of Anopheles gambiae complex using sweep regions (39 971 SNPs)', fontsize=18, fontweight='bold')\n",
    "plt.xlabel(f'PC{x} ({pcoa_results.proportion_explained[x-1]*100:.2f}% Variance)', fontsize=16, fontweight='bold')\n",
    "plt.ylabel(f'PC{y} ({pcoa_results.proportion_explained[y-1]*100:.2f}% Variance)', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Make tick labels bold\n",
    "plt.xticks(fontsize=14, fontweight='bold')\n",
    "plt.yticks(fontsize=14, fontweight='bold')\n",
    "\n",
    "# Make axes lines thicker\n",
    "ax = plt.gca()\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(3)\n",
    "\n",
    "# Save figure\n",
    "#plt.savefig(f'pcoa_taxon_PC{x}_vs_PC{y}_{region}.png', bbox_inches='tight', dpi=600)\n",
    "plt.savefig(f'pcoa_taxon_PC{x}_vs_PC{y}_all_region_all_taxon.png', bbox_inches='tight', dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a839186d",
   "metadata": {},
   "source": [
    "### For only Bissau molecular form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "948f5d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from matplotlib.cm import get_cmap\n",
    "import matplotlib.font_manager as font_manager\n",
    "\n",
    "# Extract population labels as list\n",
    "population_labels = pop_info['admin1_iso'].tolist()\n",
    "\n",
    "# Define components to plot\n",
    "x = 1\n",
    "y = 2\n",
    "\n",
    "# Unique populations, sorted for consistent color mapping\n",
    "unique_populations = sorted(np.unique(population_labels))\n",
    "\n",
    "# Get a colormap with enough unique colors\n",
    "cmap = get_cmap('tab20')  # 'Set3' is another good alternative for soft, distinct colors\n",
    "\n",
    "# Generate color map: map each population to a distinct color\n",
    "#color_map = {pop: cmap(i % 20) for i, pop in enumerate(unique_populations)}  # tab20 has 20 colors\n",
    "color_map = {\n",
    "    'GM-L': 'blue',\n",
    "    'GM-M': 'green',\n",
    "    'GM-W': 'orange',\n",
    "    'GM-N': 'purple',\n",
    "    'GM-U': 'red'\n",
    "}\n",
    "region_color_map = {\n",
    "    \"Basse\": \"orange\",\n",
    "    \"Kanifing\": \"blue\",\n",
    "    \"Janjanbureh\" : \"red\",\n",
    "    \"Kuntaur\" : \"yellow\",\n",
    "    \"Mansakonko\" : \"#33ECF5\",\n",
    "    \"Brikama\" : \"green\",\n",
    "    \"Kerewan\" : \"purple\"\n",
    "}\n",
    "# Plotting\n",
    "plt.figure(figsize=(12, 10))\n",
    "\n",
    "for pop in unique_populations:\n",
    "    mask = np.array(population_labels) == pop\n",
    "    plt.scatter(\n",
    "        pcoa_results.samples[f'PC{x}'][mask],\n",
    "        pcoa_results.samples[f'PC{y}'][mask],\n",
    "        label=pop,\n",
    "        edgecolor='k',\n",
    "        s=90,\n",
    "        linewidths=1.5,\n",
    "        color=color_map[pop]\n",
    "    )\n",
    "\n",
    "# Plot details\n",
    "plt.title(f'PCoA of Anopheles gambiae complex in sweep regions (16 823 SNPs)')\n",
    "plt.xlabel(f'PC{x} ({pcoa_results.proportion_explained[x-1]*100:.2f}% Variance)')\n",
    "plt.ylabel(f'PC{y} ({pcoa_results.proportion_explained[y-1]*100:.2f}% Variance)')\n",
    "plt.legend(title='Population', loc='best',   bbox_to_anchor=(1, 1), prop=font_manager.FontProperties(weight='bold', size=16))\n",
    "\n",
    "# Save and show\n",
    "plt.savefig(f'pcoa_bissau_all_region_PC{x}_vs_PC{y}.png', bbox_inches='tight', dpi=300)\n",
    "plt.savefig(f'pcoa_bissau_all_region_PC{x}_vs_PC{y}.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb4f3703",
   "metadata": {},
   "source": [
    "## Perform PCoA-UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4e8e321",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip uninstall -y umap\n",
    "!pip install -q umap-learn\n",
    "import umap.umap_ as umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c97c6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import umap\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "# Get  the pcoa resuts\n",
    "# Convert to NumPy array for UMAP\n",
    "pcoa_coords = pcoa_results.samples.iloc[:, :20]  # Select first 20 PCoA axes\n",
    "\n",
    "\n",
    "# Apply UMAP on PCoA coordinates\n",
    "\n",
    "umap_model = umap.UMAP(n_neighbors=15, min_dist=0.5, n_components=2, metric=\"euclidean\") # min_dist=0.5 for fine features on UMAP more visible\n",
    "umap_results = umap_model.fit_transform(pcoa_coords)  # Using the same genotype data (PCoA transformed)\n",
    "\n",
    "# Map the populations to your PCoA and UMAP results\n",
    "umap_df = pd.DataFrame(umap_results, columns=['UMAP1', 'UMAP2'])\n",
    "\n",
    "# Assuming you have a DataFrame `taxon_sample` that includes sample IDs and their populations\n",
    "taxon_sample\n",
    "# Create a mapping from sample IDs to population\n",
    "#sample_to_pop = dict(zip(taxon_sample['sample_id'], taxon_sample['taxon']))\n",
    "sample_to_pop = dict(zip(taxon_sample['sample_id'], taxon_sample['pop']))\n",
    "sample_ids = taxon_sample['sample_id'].values\n",
    "# Map the populations to your PCoA and UMAP results\n",
    "umap_df['population'] = [sample_to_pop[sample_id] for sample_id in sample_ids]  # same for umap_df\n",
    "#\n",
    "# Ensure population column has no NaN values (to avoid issues)\n",
    "umap_df = umap_df.dropna(subset=['population'])\n",
    "\n",
    "# Define custom colors for populations\n",
    "pop_color_map = {\n",
    "    'coluzzii': 'orange',\n",
    "    'bissau': 'purple',\n",
    "    'gambiae': 'blue',\n",
    "    'arabiensis': 'green',\n",
    "    'melas': 'red'\n",
    "}\n",
    "\n",
    "# Map population values to colors\n",
    "umap_df['color'] = umap_df['population'].map(color_map )\n",
    "\n",
    "# 🎨 Scatter plot using explicit colors\n",
    "plt.figure(figsize=(10, 8))\n",
    "sns.scatterplot(x=\"UMAP1\", y=\"UMAP2\", hue=\"population\", palette=color_map, data=umap_df, alpha=0.8)\n",
    "\n",
    "plt.xlabel(\"UMAP 1\")\n",
    "plt.ylabel(\"UMAP 2\")\n",
    "plt.title(f\"PCoA-UMAP in sweep region\")\n",
    "\n",
    "plt.legend(title=\"Population\")\n",
    "plt.savefig(f'pcoa_umap_PC1_vs_PC2_all_region.png')\n",
    "plt.show()\n",
    "# save the fig\n",
    "#Save umpa\n",
    "umap_df.to_csv(f'umap_taxon__PC1_vs_PC2_all_region.csv', sep=',', index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bffa218",
   "metadata": {},
   "source": [
    "## Perform UMAP on genotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cdccacf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Apply UMAP on Genotype data\n",
    "\n",
    "umap_model = umap.UMAP(n_neighbors=15, min_dist=0.5, n_components=2, metric=\"euclidean\") \n",
    "umap_results = umap_model.fit_transform(geno_alt.T)  # Using the same genotype data (PCoA transformed)\n",
    "\n",
    "# Map the populations to your PCoA and UMAP results\n",
    "umap_df = pd.DataFrame(umap_results, columns=['UMAP1', 'UMAP2'])\n",
    "\n",
    "# Assuming you have a DataFrame `taxon_sample` that includes sample IDs and their populations\n",
    "# Create a mapping from sample IDs to population\n",
    "sample_to_pop = dict(zip(taxon_sample['sample_id'], taxon_sample['taxon']))\n",
    "#sample_to_pop = dict(zip(taxon_sample['sample_id'], taxon_sample['pop']))\n",
    "sample_ids = taxon_sample['sample_id'].values\n",
    "# Map the populations to your PCoA and UMAP results\n",
    "umap_df['population'] = [sample_to_pop[sample_id] for sample_id in sample_ids]  # same for umap_df\n",
    "umap_df['sample_id'] = sample_ids\n",
    "\n",
    "# Ensure population column has no NaN values (to avoid issues)\n",
    "umap_df = umap_df.dropna(subset=['population'])\n",
    "umap_df['population'] = umap_df['population'].replace(\n",
    "    {\n",
    "        \"coluzzii\": \"An. coluzzii\",\n",
    "        \"gambiae\": \"An. gambiae s.s\",\n",
    "        \"bissau\": \"Bissau\",\n",
    "        \"arabiensis\": \"An. arabiensis\",\n",
    "        \"melas\": \"An. melas\"\n",
    "    }\n",
    ")\n",
    "# save umap_df\n",
    "umap_df.to_csv(f'umap_taxon_all_region_PC1_vs_PC2_all_taxon.csv', sep=',', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "746ed632",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare figure\n",
    "plt.figure(figsize=(12, 10))\n",
    "\n",
    "# Get unique populations\n",
    "unique_pops = umap_df['population'].unique()\n",
    "\n",
    "# Define custom colors for populations\n",
    "pop_color_map = {\n",
    "    'An. coluzzii': 'orange',\n",
    "    'Bissau': 'purple',\n",
    "    'An. gambiae s.s': 'blue',\n",
    "    'An. arabiensis': 'green',\n",
    "    'An. melas': 'red'\n",
    "}\n",
    "\n",
    "# Plot each population with white edge (using matplotlib directly)\n",
    "for pop in unique_pops:\n",
    "    subset = umap_df[umap_df['population'] == pop]\n",
    "    plt.scatter(\n",
    "        subset['UMAP1'],\n",
    "        subset['UMAP2'],\n",
    "        label=pop,\n",
    "        color=pop_color_map[pop],\n",
    "        edgecolor='white',       # White outline\n",
    "        s=200,\n",
    "        linewidths=1.5\n",
    "    )\n",
    "\n",
    "# Customize legend\n",
    "legend = plt.legend(title='Population', fontsize=21, bbox_to_anchor=(1.45, 1))\n",
    "plt.setp(legend.get_title(), fontweight='bold')\n",
    "\n",
    "for text in legend.get_texts():\n",
    "    pop_name = text.get_text()\n",
    "    text.set_fontweight('bold')\n",
    "    text.set_fontsize(22)\n",
    "    if pop_name != 'Bissau':\n",
    "        text.set_fontstyle('italic')\n",
    "\n",
    "# Bold axis labels\n",
    "plt.xlabel(\"UMAP 1\", fontsize=16, fontweight='bold')\n",
    "plt.ylabel(\"UMAP 2\", fontsize=16, fontweight='bold')\n",
    "\n",
    "# Bold tick labels\n",
    "plt.xticks(fontsize=14, fontweight='bold')\n",
    "plt.yticks(fontsize=14, fontweight='bold')\n",
    "\n",
    "# Thicker axis lines\n",
    "ax = plt.gca()\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_linewidth(3)\n",
    "\n",
    "# Title\n",
    "plt.title(f\"UMAP Projection in sweep regions\", fontsize=18, fontweight='bold')\n",
    "\n",
    "# Save\n",
    "plt.savefig(f'umap_PC1_vs_PC2_all_regions_all_taxon.png', bbox_inches='tight', dpi=600)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18d99e75",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  dowload all results\n",
    "from google.colab import files\n",
    "import glob\n",
    "# collect all png files\n",
    "png_files = glob.glob('*.png')\n",
    "csv_files = glob.glob('*.csv')\n",
    "# download all png files\n",
    "for file in png_files:\n",
    "    files.download(file)\n",
    "for file in csv_files:\n",
    "    files.download(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae8c1160",
   "metadata": {},
   "source": [
    "## FST analysis from filtered genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c4d0d31",
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_populations(pop, population_column):\n",
    "  \"\"\"\n",
    "  function that select SNP in a given region\n",
    "  \"\"\"\n",
    "  pop_samples = taxon_sample[taxon_sample[population_column] == pop]['sample_id'].tolist()\n",
    "  # Get a list of all sample IDs in your data\n",
    "  all_samples = taxon_sample['sample_id'].tolist()\n",
    "  index_pop = [all_samples.index(i) for i in pop_samples if i in all_samples]  # Check if the sample ID exists\n",
    "  return index_pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e65f861",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_label = list(taxon_sample['pop'].unique())\n",
    "pop_label\n",
    "\n",
    "pop_index = {}\n",
    "for pop in pop_label:\n",
    "  pop_index[pop] = select_populations(pop, 'pop')\n",
    "pop_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0e0a601",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fst(fst_df, region):\n",
    "    # Convert 'fst' column to numeric\n",
    "    fst_df['fst'] = pd.to_numeric(fst_df['fst'], errors='coerce')\n",
    "\n",
    "    # Define taxa order\n",
    "    order = sorted(pop_label, key=lambda x: x.split('_')[-1])\n",
    "\n",
    "    # Pivot table and reorder rows/columns\n",
    "    df = fst_df.pivot(index='population1', columns='population2', values='fst')\n",
    "    df = df.reindex(index=order, columns=order)\n",
    "\n",
    "    # Fill missing values to ensure symmetry\n",
    "    df = df.fillna(df.T)\n",
    "\n",
    "    # Generate a mask for the upper triangle\n",
    "    mask = np.triu(np.ones_like(df, dtype=bool))\n",
    "\n",
    "    # Set up the matplotlib figure\n",
    "    plt.figure(figsize=(10, 10))\n",
    "\n",
    "    # Define the colormap\n",
    "    cmap = sns.color_palette(\"Spectral_r\", n_colors=100, as_cmap=True)\n",
    "\n",
    "    # Draw the heatmap\n",
    "    sns.heatmap(\n",
    "        df, mask=mask, cmap=cmap, annot=True, fmt=\".3f\", annot_kws={\"size\": 10},\n",
    "        square=True, cbar_kws={\"shrink\": .8}, linewidths=0.5\n",
    "    )\n",
    "\n",
    "    # Adjust plot\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.yticks(rotation=0)\n",
    "    plt.title(f\"FST in Anopheles gambiae complex in {region}\", fontsize=12)\n",
    "\n",
    "    # Save and show the plot\n",
    "    plt.savefig(f\"Fst_heatmap_{region}.png\", dpi=300, bbox_inches='tight')\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ebf5caf",
   "metadata": {},
   "source": [
    "### Fst by cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c54638c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "import numpy as np\n",
    "\n",
    "def fst_processing_by_cohort(region, sample_query, filter):\n",
    "    # Step 1: Get SNP data\n",
    "    genotype_data, gm_snps_selected, taxon_sample = get_snp_data(region, sample_query, filter)\n",
    "    samples = list(taxon_sample[\"sample_id\"].values)\n",
    "\n",
    "    # Step 2: Define population indices\n",
    "    # Group samples by population\n",
    "    pop_label = list(taxon_sample['pop'].unique())\n",
    "\n",
    "    pop_index = {}\n",
    "    for pop in pop_label:\n",
    "      pop_index[pop] = select_populations(pop, 'pop')\n",
    "    pop_index\n",
    "\n",
    "    # Select genotype data for each population\n",
    "    genotype_pop = {}\n",
    "    for pop, index in pop_index.items():\n",
    "        genotype_pop[pop] = genotype_data[:, index]\n",
    "\n",
    "    # Step 3: Compute allele counts\n",
    "    allele_count = {}\n",
    "    for pop, gt in genotype_pop.items():\n",
    "        allele_count[pop] = gt.count_alleles(max_allele=3).compute()\n",
    "\n",
    "    # Step 4: Compute block length for jackknife\n",
    "    n_jack = 200  # Default value from MalariaGEN\n",
    "    n_sites = allele_count[next(iter(allele_count))].shape[0]  # Get number of SNPs from first pop\n",
    "    block_length = n_sites // n_jack  # Sites per block\n",
    "\n",
    "    # Step 5: Compute average Fst\n",
    "    fst_values = []\n",
    "    pop_names = list(pop_index.keys())\n",
    "\n",
    "    # Compute pairwise Fst for each population pair\n",
    "    for i in range(len(pop_names)):\n",
    "        for j in range(i+1, len(pop_names)):\n",
    "            pop1, pop2 = pop_names[i], pop_names[j]\n",
    "            ac1, ac2 = allele_count[pop1], allele_count[pop2]\n",
    "            # Calculate block length for jackknife.\n",
    "            n_jack = 200 # I used default value as in malariagen_data\n",
    "            n_sites = ac1.shape[0]  # number of sites\n",
    "            block_length = n_sites // n_jack  # number of sites in each block\n",
    "            # Compute blockwise Hudson's Fst\n",
    "            fst, se, _, _ = allel.blockwise_hudson_fst(ac1, ac2, blen=block_length)\n",
    "\n",
    "\n",
    "            # Store result\n",
    "            fst_values.append((pop1, pop2, float(fst), float(se)))\n",
    "    #store in a dataframe\n",
    "    fst_df = pd.DataFrame(fst_values, columns=['population1', 'population2', 'fst', 'se'])\n",
    "    #return fst_df\n",
    "    plot_fst(fst_df, region)\n",
    "\n",
    "for region in regions:\n",
    "  fst_processing_by_cohort(region, sample_query, filter)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c40c6862",
   "metadata": {},
   "source": [
    "### Fst by taxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08d5d4bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step1: get SNP data gt_filtered, gm_snps_selected, taxon_sample\n",
    "regions = [\"2R:28,430,000-28,615,000\", \"3R:28,400,000-28,600,000\",  \"X:15,130,000-15,324,000\", \"3R:32,000,000-32,080,000\"]\n",
    "#region = \"3R:32,000,000-32,080,000\"\n",
    "filter=\"gamb_colu\"\n",
    "sample_query = \"country=='Gambia, The' and taxon in ['coluzzii', 'bissau', 'gambiae']\"\n",
    "def fst_processing(region, sample_query, filter):\n",
    "\n",
    "    genotype_data, gm_snps_selected, taxon_sample = get_snp_data(region, sample_query, filter)\n",
    "    samples = list(taxon_sample[\"sample_id\"].values)\n",
    "\n",
    "\n",
    "    # Step2: construct genotype for each population by firt located sample index pop\n",
    "    # Define populations index\n",
    "    population_column = \"taxon\"\n",
    "    # gambiae index\n",
    "    gambiae_index = select_populations('gambiae', population_column)\n",
    "    # coluzzii index\n",
    "    coluzzii_index = select_populations('coluzzii', population_column)\n",
    "    # bisau index\n",
    "    bissau_index = select_populations('bissau', population_column)\n",
    "\n",
    "    # Select genotype data for each population\n",
    "    gambiae_gt = genotype_data[:, gambiae_index]\n",
    "    coluzzii_gt = genotype_data[:, coluzzii_index]\n",
    "    bissau_gt = genotype_data[:, bissau_index]\n",
    "\n",
    "    # Step3: compute allel count for each population geno_alt.count_alleles\n",
    "    gambiae_ac = gambiae_gt.count_alleles(max_allele=3).compute()\n",
    "    coluzzii_ac = coluzzii_gt.count_alleles(max_allele=3).compute()\n",
    "    bissau_ac = bissau_gt.count_alleles(max_allele=3).compute()\n",
    "    # Step4: compute block length for jackknife.\n",
    "    # Calculate block length for jackknife.\n",
    "    n_jack = 200 # I used default value as in malariagen_data\n",
    "    n_sites = gambiae_ac.shape[0]  # number of sites\n",
    "    block_length = n_sites // n_jack  # number of sites in each block\n",
    "\n",
    "\n",
    "    # Step5: compute average Fst\n",
    "\n",
    "    # Calculate average Fst.\n",
    "    fst1, se1, _, _ = allel.blockwise_hudson_fst(gambiae_ac, coluzzii_ac, blen=block_length)\n",
    "\n",
    "    # Normalise to Python scalar types.\n",
    "    fst1 = float(fst1)\n",
    "    se1 = float(se1)\n",
    "\n",
    "    fst2, se2, _, _ = allel.blockwise_hudson_fst(gambiae_ac, bissau_ac, blen=block_length)\n",
    "\n",
    "    fst2 = float(fst2)\n",
    "    se2 = float(se2)\n",
    "\n",
    "    fst3, se3, _, _ = allel.blockwise_hudson_fst(coluzzii_ac, bissau_ac, blen=block_length)\n",
    "\n",
    "    fst3 = float(fst3)\n",
    "    se3 = float(se3)\n",
    "\n",
    "    # store result into a dataframe\n",
    "    fst_df = pd.DataFrame(columns=['fst', 'se', 'population1', 'population2'])\n",
    "    fst_df.loc[0] = [fst1, se1, 'gambiae', 'coluzzii']\n",
    "    fst_df.loc[1] = [fst2, se2, 'gambiae', 'bissau']\n",
    "    fst_df.loc[2] = [fst3, se3, 'coluzzii', 'bissau']\n",
    "\n",
    "    # reformating\n",
    "    fst_df['population1'] = fst_df['population1'].replace({\n",
    "        'gambiae': 'An. gambiae s.s',\n",
    "        'coluzzii': 'An. coluzzii',\n",
    "        'bissau': 'Bissau'\n",
    "    })\n",
    "    fst_df['population2'] = fst_df['population2'].replace({\n",
    "        'gambiae': 'An. gambiae s.s',\n",
    "        'coluzzii': 'An. coluzzii',\n",
    "        'bissau': 'Bissau'\n",
    "    })\n",
    "    fst_df\n",
    "\n",
    "    # Convert 'fst' column to numeric\n",
    "    fst_df['fst'] = pd.to_numeric(fst_df['fst'], errors='coerce')\n",
    "\n",
    "    # Define taxa order\n",
    "    order = ['An. gambiae s.s', 'An. coluzzii', 'Bissau']  # Adjust according to your data\n",
    "\n",
    "    # Pivot table and reorder rows/columns\n",
    "    df = fst_df.pivot(index='population1', columns='population2', values='fst')\n",
    "    df = df.reindex(index=order, columns=order)\n",
    "\n",
    "    # Fill missing values to ensure symmetry\n",
    "    df = df.fillna(df.T)\n",
    "\n",
    "    # Generate a mask for the upper triangle\n",
    "    mask = np.triu(np.ones_like(df, dtype=bool))\n",
    "\n",
    "    # Set up the matplotlib figure\n",
    "    plt.figure(figsize=(5, 5))\n",
    "\n",
    "    # Define the colormap\n",
    "    cmap = sns.color_palette(\"Spectral_r\", n_colors=100, as_cmap=True)\n",
    "\n",
    "    # Draw the heatmap\n",
    "    sns.heatmap(\n",
    "        df, mask=mask, cmap=cmap, annot=True, fmt=\".3f\", annot_kws={\"size\": 10},\n",
    "        square=True, cbar_kws={\"shrink\": .8}, linewidths=0.5\n",
    "    )\n",
    "\n",
    "    # Adjust plot\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.yticks(rotation=0)\n",
    "    plt.title(f\"FST in Anopheles gambiae complex in {region}\", fontsize=12)\n",
    "\n",
    "    # Save and show the plot\n",
    "    plt.savefig(f\"Fst_heatmap_{region}.png\", dpi=300, bbox_inches='tight')\n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
