{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "efd71d2e",
   "metadata": {},
   "source": [
    "# GWSS analysis using Fst_gwss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c753eaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -qq malariagen_data\n",
    "!pip install -q petl\n",
    "\n",
    "\n",
    "import malariagen_data\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import allel\n",
    "import zarr\n",
    "import dask\n",
    "import dask.array as da\n",
    "# silence some dask warnings\n",
    "dask.config.set(**{'array.slicing.split_large_chunks': True})\n",
    "from dask.diagnostics.progress import ProgressBar\n",
    "\n",
    "import random\n",
    "import functools\n",
    "import petl as ptl\n",
    "import itertools\n",
    "import scipy\n",
    "from collections import Counter\n",
    "\n",
    "# plotting setup\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import matplotlib_venn as venn\n",
    "import seaborn as sns\n",
    "\n",
    "ag3 = malariagen_data.Ag3()\n",
    "ag3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cfcad34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# step1: Define population: between An. gambia complex or withing each taxon\n",
    "taxons = [\"coluzzii\", \"gambiae\", \"bissau\"]\n",
    "taxon=\"bissau\"\n",
    "country = \"Gambia, The\"\n",
    "sample_query = f\"country=='{country}' and taxon=='{taxon}'\"\n",
    "\n",
    "taxon_sample = ag3.sample_metadata(sample_query=sample_query)\n",
    "#taxon_sample['pop'] = taxon_sample['admin1_iso'] + '_' + taxon_sample['taxon']\n",
    "#taxon_sample.groupby('pop').size()\n",
    "#taxon_sample = taxon_sample[taxon_sample['pop'].isin(taxon_sample['pop'].value_counts()[taxon_sample['pop'].value_counts() >= 10].index)]\n",
    "\n",
    "# Get 100 samples from coluzzii\n",
    "#coluzzii_subset = taxon_sample.query(\"taxon=='coluzzii'\").groupby(\"admin1_iso\").apply(lambda x: x.sample(n=min(20, len(x)))).reset_index(drop=True)\n",
    "\n",
    "# Get all other taxons (gambiae and bissau)\n",
    "#other_taxa_sample = taxon_sample[taxon_sample[\"taxon\"] != \"coluzzii\"]\n",
    "\n",
    "# Concatenate both to redefine taxon_sample\n",
    "#taxon_sample = pd.concat([coluzzii_subset, other_taxa_sample], ignore_index=True)\n",
    "\n",
    "\n",
    "#gambiae_subset_index = taxon_sample.query(\"taxon=='gambiae'\").index\n",
    "#coluzzii_subset_index = taxon_sample.query(\"taxon=='coluzzii'\").index\n",
    "#bissau_subset_index = taxon_sample.query(\"taxon=='bissau'\").index\n",
    "\"\"\"taxon_pop = {\n",
    "  'gambiae': gambiae_subset_index,\n",
    "  'coluzzii': coluzzii_subset_index,\n",
    "  'bissau': bissau_subset_index\n",
    "}\"\"\"\n",
    "#taxon_pop = dict([(f\"{p}\", list(df.index)) for p, df in taxon_sample.reset_index().groupby(['taxon'])])\n",
    "taxon_sample['pop'] = taxon_sample['admin1_iso'] + '_' + taxon_sample['taxon']\n",
    "#cohort_pop = dict([(f\"{p}\", list(df.index)) for p, df in taxon_sample.reset_index().groupby(['pop'])])\n",
    "pops = taxon_sample['pop'].unique()\n",
    "taxon_pop = {pop: taxon_sample[taxon_sample['pop'] == pop].index for pop in pops}\n",
    "taxon_pop.keys()\n",
    "#pops = taxon_sample['taxon'].unique()\n",
    "\n",
    "\n",
    "# step2: Get SNP data\n",
    "samples_list = list(taxon_sample['sample_id'])\n",
    "sample_query = f\"sample_id in {samples_list}\"\n",
    "#To access to the genotypes within the 2L chromosomes\n",
    "chrom2L_gt = ag3.snp_calls(region=\"3R\", sample_query=sample_query)\n",
    "\n",
    "# To filter the SNP dataset and warp the dataset to GT array\n",
    "filt = 'gamb_colu'\n",
    "filt_val = chrom2L_gt[f\"variant_filter_pass_{filt}\"].values\n",
    "gt_filtered = allel.GenotypeDaskArray(chrom2L_gt[\"call_genotype\"][filt_val].data)\n",
    "filterd_val = chrom2L_gt[f\"variant_filter_pass_{filt}\"].values\n",
    "filtered_pos = allel.SortedIndex(chrom2L_gt['variant_position'][filterd_val].values)\n",
    "\n",
    "gt_filtered.shape #(37199402, 351, 2)\n",
    "# step3: Compute allel count for each population for a given chromosome\n",
    "\n",
    "ac_pop_taxon = {} ## compute allele counts pop - dict\n",
    "with ProgressBar():\n",
    "  for pop in taxon_pop.keys():\n",
    "    print(f\"Computing allele count for {pop}\")\n",
    "    ac_pop_taxon[pop] = gt_filtered.take(taxon_pop[pop],axis=1).count_alleles(max_allele=3).compute()\n",
    "\n",
    "# step4: Get SNP position then sort it.\n",
    "#snps_pos = allel.SortedIndex(chrom2L_gt['variant_position'].values)\n",
    "#filtered_pos = allel.SortedIndex(chrom2L_gt['variant_position'][filterd_val].values)\n",
    "filtered_pos.shape\n",
    "\n",
    "# Save the ac_pop_taxon dict\n",
    "import pickle\n",
    "with open('ac_pop_taxon.pkl', 'wb') as f:\n",
    "    pickle.dump(ac_pop_taxon, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4e953ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache(maxsize=None)\n",
    "def load_ac(pop):\n",
    "    ac_popi = ac_pop_taxon[pop]\n",
    "    return ac_popi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71146a55",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_windowed_pair_fst(pops, filt_pos, size):\n",
    "    # order is irrelevant\n",
    "    pop1, pop2 = pops\n",
    "    ac_pop1 = load_ac(pop=pop1)\n",
    "    ac_pop2 = load_ac(pop=pop2)\n",
    "    loc_pass = ac_pop1.is_segregating() & ac_pop2.is_segregating() & (ac_pop1.max_allele() <= 1) & (ac_pop2.max_allele() <= 1)\n",
    "    # Find selected SNPs pos\n",
    "    pos=filt_pos[loc_pass]\n",
    "    # compute windowed_fst\n",
    "    fst_wind, windows, counts = allel.windowed_hudson_fst(pos, ac_pop1[loc_pass], ac_pop2[loc_pass],size=size)\n",
    "    return fst_wind, windows, counts\n",
    "\n",
    "def compute_windowed_pairwise_fst(coh_list, pos_filt, size):\n",
    "    pairwise_fst = {}\n",
    "    wind_pairwise_fst = {}\n",
    "    for pop1, pop2 in itertools.combinations(coh_list, 2):\n",
    "        pops = tuple(sorted([pop1, pop2]))\n",
    "        fst, wind, _ = compute_windowed_pair_fst(pops=pops,filt_pos=pos_filt,size=size)\n",
    "        pairwise_fst[f'fst_{pop1}_{pop2}'] = fst\n",
    "        wind_pairwise_fst[f'wind_{pop1}_{pop2}'] = wind\n",
    "    return pairwise_fst, wind_pairwise_fst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d113bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute fst and windows\n",
    "import json\n",
    "fst_dict_ag, wind_fst_dict_ag = compute_windowed_pairwise_fst(coh_list=pops, pos_filt=filtered_pos, size=100000)\n",
    "type(wind_fst_dict_ag)\n",
    "# save a dict\n",
    "fst_dict_ag_cpy = fst_dict_ag.copy()\n",
    "for key in fst_dict_ag_cpy:\n",
    "    fst_dict_ag_cpy[key] = fst_dict_ag_cpy[key].tolist()  # Convert each ndarray to a list\n",
    "\n",
    "# save a dict\n",
    "with open('fst_gwss_taxon.json', 'w') as f:\n",
    "    json.dump(fst_dict_ag_cpy, f)\n",
    "\n",
    "# warp fst data to dataframe\n",
    "df_fst=pd.DataFrame(fst_dict_ag)\n",
    "df_fst\n",
    "df_fst.to_csv('fst_gwss_taxon.csv')\n",
    "# load data\n",
    "#df_fst = pd.read_csv('/content/fst_gwss_taxon.csv')\n",
    "# Extract windowed FST means for all population pairs\n",
    "\"\"\"df_wind = pd.DataFrame({\n",
    "    key: np.mean(wind_fst_dict_ag[key], axis=1)  # Compute mean FST for each pair\n",
    "    for key in wind_fst_dict_ag.keys()\n",
    "})\"\"\"\n",
    " # Rename columns to indicate pairs\n",
    "#df_wind.columns = [key.replace('wind_', '') for key in wind_fst_dict_ag.keys()]\n",
    "key = list(wind_fst_dict_ag.keys())[0]\n",
    "df_wind = pd.DataFrame(np.mean(wind_fst_dict_ag[key], axis=1), columns=['wind_mean'])\n",
    "df_wind.to_csv('wind_mean_3R.csv')\n",
    "data_fst_ag = pd.concat([df_wind, df_fst], axis=1)\n",
    "ag_fst = data_fst_ag.copy()\n",
    "ag_fst[ag_fst < 0] = 0  # Ensure FST values are non-negative\n",
    "ag_fst.to_csv('fst_gwss_bissau_3R.csv')\n",
    "\n",
    "\n",
    "\n",
    "# Warp to df\n",
    "data_fst_ag = ag_fst.melt('wind_mean', var_name='taxon_Pop', value_name='pair_fst')\n",
    "data_fst_ag.to_csv('fst_gwss_bissau_3R.csv')\n",
    "#data_fst_ag = data_fst_ag.query(\"taxon_Pop =='fst_coluzzii_gambiae'\")\n",
    "xlim1 = list(ag_fst.wind_mean)\n",
    "\n",
    "#data_fst_ag.query(\"pair_fst >= 0.2 and taxon_Pop != 'fst_gambia_bissau' and 6000000 <= wind_mean <= 10000000\").sort_values(by='wind_mean', ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5920e612",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "def plot_fst_gwss(data=data_fst_ag, figsize=(12,4), ax=None, title=None, xlab=None,\n",
    "                  ylab=None, legend_loc=None, fig_name=None):\n",
    "\n",
    "    if ax is None:\n",
    "        fig, ax = plt.subplots(figsize=figsize)\n",
    "        sns.despine(ax=ax, offset=5)\n",
    "\n",
    "    # Plot FST\n",
    "    sns.lineplot(x='wind_mean', y='pair_fst', hue='taxon_Pop', data=data, ax=ax, linewidth=0.5)\n",
    "\n",
    "    # Set axis titles\n",
    "    if title:\n",
    "        ax.set_title(title)\n",
    "    if xlab:\n",
    "        ax.set_xlabel(xlab)\n",
    "    if ylab:\n",
    "        ax.set_ylabel(ylab)\n",
    "\n",
    "    # Set legend\n",
    "    if legend_loc:\n",
    "        ax.legend(loc='best', bbox_to_anchor=legend_loc)\n",
    "\n",
    "    # Set axis limits\n",
    "    xlim = [xlim1[0], xlim1[-1]]\n",
    "    ax.set_xlim(*xlim)\n",
    "    #ax.set_ylim(0, 1)  # Change y-axis range from 0 to 1\n",
    "    ax.set_ylim(0,0.25)\n",
    "    #ax.set_yticks([i / 20 for i in range(21)])  # Set ticks at 0.05 intervals (0, 0.05, 0.10, ..., 1.0)\n",
    "    ax.set_yticks([i / 10 for i in range(11)])\n",
    "\n",
    "    # Format x-axis tick labels\n",
    "    ax.set_xticklabels(['{:,}'.format(int(v)) for v in ax.get_xticks()])\n",
    "\n",
    "    # Save figure\n",
    "    if fig_name:\n",
    "        fig.savefig(f\"{fig_name}.png\", dpi=300, bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1741e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compute An. gambiae Fst\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(12, 4))\n",
    "sns.despine(ax=ax, offset=5)\n",
    "plot_fst_gwss(data=data_fst_ag, ax=ax, xlab='Chrom 3R (bp)', ylab='Pairwise $F_{ST}$ in 100kb windows',\n",
    "              legend_loc=(1, 1))\n",
    "\"\"\"\n",
    "ax.annotate('$vgsc$ region', xy=(2431617, 0.7), xytext=(600000, 1),\n",
    "            color='darkred', arrowprops=dict(arrowstyle=\"->\", color='slategrey'))\n",
    "ax.annotate('$OBP41$ region', xy=(11551855.5, 0.3), xytext=(15000000, 0.5),\n",
    "            color='darkred', arrowprops=dict(arrowstyle=\"->\", color='slategrey'))\n",
    "ax.annotate('$Ors$ region', xy=(32080000, 0.05), xytext=(31000000, 0.10),\n",
    "            color='darkred', arrowprops=dict(arrowstyle=\"->\", color='slategrey'))\"\"\"\n",
    "\n",
    "# Save fig\n",
    "fig.savefig(f\"3R_window_bissau_fst.png\", dpi=300, bbox_inches='tight')\n",
    "fig.show()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
